lib/encoder.js

```diff
@@ -908,70 +955,130 @@ function defineInstanceMembers() {
+        offset += elemLength;
+      }
     }
-    return new Float32Array(rv);
+    if (offset !== buffer.length) {
+      throw new TypeError('Extra bytes found after decoding the vector');
+    }
+    const typeInfo = types.getDataTypeNameByCode(params.info);
+    return new Vector(rv, typeInfo);
   };
 
   /**
-   * @param {CqlVector} value
-   * @param {Object} params
+   * @param {ColumnInfo} cqlType
+   * @returns {Number}
    */
-  this.encodeVector = function(value, params) {
-
-    // Evaluate params to encodeVector(), returning the computed subtype
-    function evalParams() {
-
-      if (!(value instanceof Float32Array)) {
-        throw new TypeError("Driver only supports vectors of 4 byte floating point values");
-      }
-
-      // Perform client-side validation iff we were actually supplied with meaningful type info.  In practice
-      // this will only occur when using prepared statements.
-      if (params.hasOwnProperty("subtype") && params.hasOwnProperty("dimensions")) {
-
-        const subtype = params["subtype"];
-        const dimensions = params["dimensions"];
-        if (value.length !== dimensions) {
-          throw new TypeError(`Expected vector with ${dimensions} dimensions, observed size of ${value.length}`);
-        }
-        if (subtype.code !== dataTypes.float) {
-          throw new TypeError("Driver only supports vectors of 4 byte floating point values");
+  this.serializationSizeIfFixed = function (cqlType) {
+    switch (cqlType.code) {
+      case dataTypes.bigint:
+        return 8;
+      case dataTypes.boolean:
+        return 1;
+      case dataTypes.timestamp:
+        return 8;
+      case dataTypes.double:
+        return 8;
+      case dataTypes.float:
+        return 4;
+      case dataTypes.int:
+        return 4;
+      case dataTypes.timeuuid:
+        return 16;
+      case dataTypes.uuid:
+        return 16;
+      
+      case dataTypes.ascii:
+      case dataTypes.tinyint:
+      case dataTypes.blob:
+      case dataTypes.counter:
+      case dataTypes.decimal:
+      case dataTypes.duration:
+      case dataTypes.inet:
+      case dataTypes.varint:
+      case dataTypes.smallint:
+      case dataTypes.date:
+      case dataTypes.time:
+      case dataTypes.text:
+      case dataTypes.varchar:
+      case dataTypes.list:
+      case dataTypes.map:
+      case dataTypes.set:
+      case dataTypes.udt:
+      case dataTypes.tuple:
+        return -1;
+      case dataTypes.custom:
+        if ('customTypeName' in cqlType && cqlType.customTypeName === 'vector'){ 
+          const subtypeSerialSize = this.serializationSizeIfFixed(cqlType.info);
+          if (subtypeSerialSize === -1){
+            return -1;
+          }
+          return subtypeSerialSize * cqlType.dimension;
+          
         }
-        return subtype;
-      }
+        return -1;
+      default:
+        throw new TypeError('Cannot calculate size');
```

Comment:
Hmmmm... I'm wondering if it's better to make the default case return -1.  You wouldn't have to specify all the data types that _should_ have this behaviour (like you do in this impl)... you'd just specify size functions for the types that _are_ of a fixed size and everybody else returns -1.

Best argument I can think of _not_ to do it that way is that there may be some benefit in being explicit for every single type: is it fixed size or not?  It does make the code a bit more cumbersome, though.